File: react-native/src/data/repositories/SymptomRepositoryImpl.ts
import { SymptomEntry } from '../../domain/entities/SymptomEntry';
import { ISymptomRepository } from '../../domain/repositories/ISymptomRepository';
import { SymptomLocalDataSource } from '../datasources/SymptomLocalDataSource';

/**
 * Implementação do repositório de sintomas.
 *
 * Conecta a interface do domínio (ISymptomRepository) ao data source
 * concreto (SymptomLocalDataSource). Segue o princípio de inversão
 * de dependência: o domínio define o contrato, e esta camada o implementa.
 *
 * Ponto de extensão: poderia adicionar cache em memória, sincronização
 * com API remota, ou estratégias de retry sem alterar o domínio.
 */
export class SymptomRepositoryImpl implements ISymptomRepository {
  constructor(private readonly localDataSource: SymptomLocalDataSource) {}

  async save(entry: SymptomEntry): Promise<void> {
    await this.localDataSource.save(entry);
  }

  async getAll(): Promise<SymptomEntry[]> {
    return this.localDataSource.getAll();
  }

  async getById(id: string): Promise<SymptomEntry | null> {
    return this.localDataSource.getById(id);
  }

  async delete(id: string): Promise<void> {
    await this.localDataSource.delete(id);
  }
}

File: kotlin/app/src/main/java/com/vitalcheck/app/data/repository/SymptomRepositoryImpl.kt
package com.vitalcheck.app.data.repository

import com.vitalcheck.app.data.datasource.SymptomLocalDataSource
import com.vitalcheck.app.domain.entity.SymptomEntry
import com.vitalcheck.app.domain.repository.ISymptomRepository

/**
 * Implementação do repositório de sintomas.
 *
 * Delega ao SymptomLocalDataSource (Room).
 * Abstração permite futura adição de cache em memória
 * ou sincronização com servidor remoto.
 */
class SymptomRepositoryImpl(
    private val localDataSource: SymptomLocalDataSource
) : ISymptomRepository {

    override suspend fun save(entry: SymptomEntry) =
        localDataSource.save(entry)

    override suspend fun getAll(): List<SymptomEntry> =
        localDataSource.getAll()

    override suspend fun deleteById(id: String) =
        localDataSource.deleteById(id)
}

